### PR TITLE
Build fixes and a small refactor

### PR DESCRIPTION
**Pull request by karlri**
Really awesome tool, thank you! Now it works in proxy environment and with async-std. I also refactored one file with unsafe code that was correct but now that MaybeUninit landed it's perhaps more appropriate to use that. The refactor enforces the C lifetime contracts rather than addressing them in comments.

### CODE REVIEW

**Comment by karlri** [comment]
I suppose it's strange to look only at http_proxy, I guess https_proxy might differ.

File: src/cli/mod.rs [hunk 1]
```diff
@@ -222,6 +222,14 @@ impl Invocation {
             "CARGO_TARGET_ARMV5TE_AXIS_LINUX_GNUEABI_RUSTFLAGS=--cfg crossbeam_no_atomic_64",
         ]);
 
+        // Pass through some env vars to docker container that cargo might need to fetch dependencies
+        let pass_through_env = ["http_proxy", "https_proxy"];
+        for var in pass_through_env {
+            if let Ok(val) = std::env::var("http_proxy") {
```
Comment (karlri) [inline src/cli/mod.rs:228]: should probably check var(var)

File: src/whoami.rs [hunk 2]
```diff
@@ -16,43 +13,69 @@ pub fn whoami() -> Whoami {
...
+    pub struct SafePasswd<'a>(&'a libc::passwd);
+
+    impl<'a> SafePasswd<'a> {
+        /// calls libc::getpwuid_r but forces params to be correct and enforces the lifetime contract.
+        /// user cant access/move/drop the buffers until return value is dropped.
+        pub fn getpwuid_r(
+            uid: u32,
+            pwd_buf: &'a mut MaybeUninit<libc::passwd>,
+            string_buf: &'a mut MaybeUninit<[c_char; STRING_BUF_LEN]>,
+        ) -> Option<SafePasswd<'a>> {
```
Comment (willglynn) [inline src/whoami.rs:36]: 👍

File: src/cli/mod.rs (hunk 1 above)
Comment (willglynn) [reply src/cli/mod.rs:228]: Hmm… checking various tools, I'm thinking that maybe we should pass through any/all upper and lowercase versions of:
```
[
  "http_proxy",
  "https_proxy",
  "ftp_proxy",
  "all_proxy",
  "no_proxy",
]
```

I have no idea which tools inside the build container will respect which variables with which precedence, but at least that way `cargo-acap` isn't making that decision.

File: src/cli/mod.rs [hunk 3]
```diff
@@ -222,6 +222,14 @@ impl Invocation {
...
+                docker.args(&["--env", &format!("{}={}", var, val)]);
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (willglynn) [reply src/cli/mod.rs:232]: `--mount=type=ssh` applies to `RUN` commands in `docker build`s, but we're running a program in an already-built container via `docker run`. We could pass through SSH context but AFAIK we'd have to do it ourselves. Maybe this is better handled through `DOCKER_OPTS`?

File: src/cli/mod.rs [hunk 4]
```diff
@@ -204,6 +204,32 @@ impl Invocation {
             &format!("{}:/.cargo:Z", self.cargo_home.display()),
         ]);
 
+        // This hack should become obsolete when this lands: https://github.com/rust-lang/rust/issues/44839
+        // One would wish that one could #[cfg(target-feature = atmoic64)] but that's not possible today.
+        // Therefore crates do lots of #[cfg(target = "mipsel-unknown-linux-gnu")] to determine 64bit atomic support or not.
```
[the remark this answers is not in the document]
Comment (karlri) [reply src/cli/mod.rs:209]: Completely agree, they're working on supporting conditional compilation for target feature supports atomic 64 ops. Then eventually when crossbeam bumps the minimum supported Rust version so they can actually use that, this hack can be removed.

File: src/cli/mod.rs [hunk 5]
```diff
@@ -204,6 +204,32 @@ impl Invocation {
...
+        // * set a RUSTFLAGS env var in docker
+        // * applies only to rustc invocations with specific target params that don't support 64 bit atomic ops.
+        // * sets a rustc cfg crossbeam_no_atomic_64
+        // * causes rustc #[cfg(crossbeam_no_atomic_64)] fallback code to compile in
+        //   https://docs.rs/crossbeam-utils/0.8.10/src/crossbeam_utils/atomic/atomic_cell.rs.html#728
+        docker.args(&[
+            "--env",
+            "CARGO_TARGET_MIPSEL_AXIS_LINUX_GNU_RUSTFLAGS=--cfg crossbeam_no_atomic_64",
+            "--env",
+            "CARGO_TARGET_ARMV5TE_AXIS_LINUX_GNUEABI_RUSTFLAGS=--cfg crossbeam_no_atomic_64",
```
[the remark this answers is not in the document]
Comment (karlri) [reply src/cli/mod.rs:222]: I agree with you. The reason it's here is I first was planning to send RUSTFLAGS conditionally but that also affects compilation of native build tools in build scripts so I had to scrap that idea. With the current solution to set rustflags only for specific targets, it's definitely better in the Dockerfile.

File: src/cli/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (karlri) [reply src/cli/mod.rs:232]: I don't know enough about docker to comment on this. Maybe we close this and perhaps handle it separately?

File: src/cli/mod.rs (hunk 1 above)
Comment (karlri) [reply src/cli/mod.rs:228]: Agreed! Will do tomorrow!

File: src/cli/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (willglynn) [reply src/cli/mod.rs:209]: Works for me!

**Comment by karlri** [comment]
I guess my changes would require publishing a new docker image, I don't know how but I built one locally. I tested with:
```
docker build --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy .
docker tag 696bb277c02a karlattempt2:1.61.0
```
Then I tested that I do get the errors without that special image:
```
cargo acap build # expected errors
[... snip ...]
error[E0432]: unresolved imports `core::sync::atomic::AtomicI64`, `core::sync::atomic::AtomicU64`
  --> /.cargo/registry/src/github.com-1ecc6299db9ec823/crossbeam-utils-0.8.10/src/lib.rs:79:49
   |
79 |             pub(crate) use core::sync::atomic::{AtomicI64, AtomicU64};
   |                                                 ^^^^^^^^^  ^^^^^^^^^ no `AtomicU64` in `sync::atomic`
   |                                                 |
   |                                                 no `AtomicI64` in `sync::atomic`
```
Then I tested that the errors go away with my image:
```
➜  your-app git:(master) ✗ cargo acap --docker-image=karlattempt2 build
cargo-acap: using Docker image karlattempt2:1.61.0
cargo-acap: building target aarch64
    Finished release [optimized] target(s) in 0.02s
cargo-acap: building target armv5tej
    Finished release [optimized] target(s) in 0.03s
cargo-acap: building target armv6
    Finished release [optimized] target(s) in 0.04s
cargo-acap: building target armv7
    Finished release [optimized] target(s) in 0.03s
cargo-acap: building target armv7hf
    Finished release [optimized] target(s) in 0.03s
cargo-acap: building target mips
    Finished release [optimized] target(s) in 0.02s
```

**Comment by willglynn** [comment]
Now, how best to rebuild existing image tags…

**Comment by willglynn** [comment]
Oh, that was straightforward. I manually invoked [the relevant GitHub action](https://github.com/trunnion/cargo-acap/actions/workflows/build-rust-image.yml) for `1.62.0`, `1.61.0`, and `1.60.0`, which should (🤞) result in new builds from the updated `Dockerfile` getting pushed to the corresponding tags.

`cargo-acap` invokes `docker run` and not `docker run --pull`, so you may have to `docker pull trunnion/cargo-acap:1.62.0` to pull the updated image with that tag once that [lands at the registry](https://hub.docker.com/r/trunnion/cargo-acap/tags).

**Pull request merged**